ui/app/src/services/configuration.ts

```diff
@@ -196,19 +197,20 @@ export function setActiveTargetingModel(data): Observable<ActiveTargetingModel>
 
 // region SidebarConfig
 
-export function getSiteUiConfig(site: string): Observable<any> {
+export function getSiteUiConfig(
+  site: string
+): Observable<Omit<GlobalState['uiConfig'], 'error' | 'isFetching' | 'currentSite'>> {
   return getConfigurationDOM(site, '/ui.xml', 'studio').pipe(
     map((xml) => {
       if (xml) {
         const widgets = xml.querySelector('[id="craftercms.components.ToolsPanel"] > configuration > widgets');
         if (widgets) {
+          widgets.querySelectorAll('widget').forEach((e, index) => e.setAttribute('uiKey', String(index)));
```

Comment:
Please add comment explaining why we had to add this:

When rendering widgets dynamically and changing pages on the tools panel, if there are duplicate react `key` props across pages, react may no swap the components correctly, incurring in unexpected behaviours. We need a unique key for each widget.

Add comment to the createwidgetdescriptor function too.